partner_invoicing_mode_ten_days/models/res_partner.py

```diff
@@ -0,0 +1,40 @@
+# Copyright 2023 ACSONE SA/NV
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from dateutil.relativedelta import relativedelta
+
+from odoo import fields, models
+
+
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+
+    invoicing_mode = fields.Selection(
+        selection_add=[("ten_days", "Each ten days")],
+        ondelete={"ten_days": "set default"},
+    )
+
+    def _update_next_invoice_date(self):
+        """
+        This will update the next invoice date from the configuration set on
+        the partner if needed (not for standard invoicing_mode).
+        """
+        today = fields.Datetime.now()
+        ten_days_partners = self.filtered(
+            lambda partner: partner.invoicing_mode == "ten_days"
+        )
+        if ten_days_partners:
+            rows = [10, 20, (today + relativedelta(day=31)).day]
+            try:
+                next_day = next(row for row in rows if today.day < row)
+                next_invoice_date = today.replace(day=next_day)
+            except StopIteration:
+                next_day = rows[0]
+                next_invoice_date = (today + relativedelta(months=+1)).replace(
+                    day=next_day
+                )
```

Comment:
Idea to avoid the try,

```suggestion
            last_month_day = today + relativedelta(day=31)
            months = 0
            day = 10
            if last_month_day == today:
                months = 1
            else:
                day = {0: 10, 1: 20, 2: 31, 3: 31}.get(today.day // 10)
            next_invoice_date = today + relativedelta(day=day, months=months)
```